dali/python/nvidia/dali/plugin/numba.py

```diff
@@ -0,0 +1,193 @@
+from nvidia.dali.pipeline import Pipeline
+from nvidia.dali import ops
+from nvidia.dali import types as dali_types
+from numba import types as numba_types
+from numba import njit, cfunc, carray
+import numpy as np
+import numba as nb
+
+_to_numpy = {
+    dali_types.UINT8 : "uint8",
+    dali_types.UINT16 : "uint16",
+    dali_types.UINT32 : "uint32",
+    dali_types.UINT64 : "uint64",
+    dali_types.INT8 : "int8",
+    dali_types.INT16 : "int16",
+    dali_types.INT32 : "int32",
+    dali_types.INT64 : "int64",
+    dali_types.FLOAT16 : "float16",
+    dali_types.FLOAT : "float32",
+    dali_types.FLOAT64 : "float64",
+}
+
+@nb.extending.intrinsic
+def address_as_void_pointer(typingctx, src):
+    from numba.core import types, cgutils
+    sig = types.voidptr(src)
+
+    def codegen(cgctx, builder, sig, args):
+        return builder.inttoptr(args[0], cgutils.voidptr_t)
+    return sig, codegen
+
+@njit
+def _get_shape_view(shapes_ptr, ndims_ptr, num_dims, num_samples):
+    ndims = carray(address_as_void_pointer(ndims_ptr), num_dims, dtype=np.int32)
+    samples = carray(address_as_void_pointer(shapes_ptr), (num_dims, num_samples), dtype=np.int64)
+    l = []
+    for sample, size in zip(samples, ndims):
+        d = []
+        for shape_ptr in sample:
+            d.append(carray(address_as_void_pointer(shape_ptr), size, dtype=np.int64))
+        l.append(d)
+    return l
+
+class NumbaFunc(ops.NumbaFunctionBase):
+    ops.register_cpu_op('NumbaFunc')
+
+    def _setup_fn_sig(self):
+        return numba_types.void(numba_types.uint64,
+                        numba_types.uint64,
+                        numba_types.int32,
+                        numba_types.uint64,
+                        numba_types.uint64,
+                        numba_types.int32, numba_types.int32)
+                
+    def _run_fn_sig(self, batch_processing=False):
+        sig_types = []
+        sig_types.append(numba_types.uint64)
+        sig_types.append(numba_types.uint64)
+        sig_types.append(numba_types.uint64)
+        sig_types.append(numba_types.int32)
+
+        sig_types.append(numba_types.uint64)
+        sig_types.append(numba_types.uint64)
+        sig_types.append(numba_types.uint64)
+        sig_types.append(numba_types.int32)
+
+        if batch_processing:
+            sig_types.append(numba_types.int32)
+        return numba_types.void(*sig_types)
+
+    def _get_carray_eval_lambda(self, dtype, ndim):
+        eval_string = "lambda ptr, shape: carray(ptr, ("
+        for i in range(ndim):
```

Comment:
I assume this loop is needed because we cannot just call `tuple(some_carray)`, right?

Reply:
Yes, we can't.